File: src/main/scala/com/github/chengpohi/edql/parser/EDQLPsiInterceptor.scala
package com.github.chengpohi.edql.parser

import com.github.chengpohi.edql.parser.json.{JsonCollection, JsonValParser}
import com.github.chengpohi.edql.parser.psi._
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.util.Computable
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.psi.{PsiComment, PsiErrorElement, PsiFile, PsiWhiteSpace}
import org.apache.commons.collections.CollectionUtils
import org.apache.commons.lang3.RandomStringUtils

import java.net.URL
import scala.jdk.CollectionConverters.ListHasAsScala
import scala.util.Try

class EDQLPsiInterceptor(val parserFactory: EDQLParserFactory) extends InterceptFunction with JsonValParser {
  import parserFactory._

  def parseJson(text: String): Try[JsonCollection.Val] = {
    ApplicationManager.getApplication.runReadAction(new Computable[Try[JsonCollection.Val]] {
      override def compute(): Try[JsonCollection.Val] = {
        try {
          val psiFile = createFile("dummmy", text)

          val element = PsiTreeUtil.findChildOfType(psiFile, classOf[PsiErrorElement])
          if (element != null) {
            throw new RuntimeException(element.getErrorDescription)
          }

          val expr = PsiTreeUtil.findChildOfType(psiFile, classOf[EDQLExpr])
          if (!expr.textMatches(text)) {
            throw new RuntimeException("parse failed: " + text)
          }
          scala.util.Success(toJsonVal(expr))
        } catch {
          case ex: Throwable => {
            scala.util.Failure(ex)
          }
        }
      }
    })
  }

  def parseExpr(text: String): Try[PsiFile] = {
    ApplicationManager.getApplication.runReadAction(new Computable[Try[PsiFile]] {
      override def compute(): Try[PsiFile] = {
        try {
          val psiFile = createFile("dummmy", text)
          val element = PsiTreeUtil.findChildOfType(psiFile, classOf[PsiErrorElement])
          if (element != null) {
            throw new RuntimeException(element.getErrorDescription)
          }

          scala.util.Success(psiFile)
        } catch {
          case ex: Exception => scala.util.Failure(ex)
        }
      }
    })
  }

  def parse(text: String): Try[Seq[Instruction2]] = {
    ApplicationManager.getApplication.runReadAction(new Computable[Try[Seq[Instruction2]]] {
      override def compute(): Try[Seq[Instruction2]] = {
        try {
          val psiFile = createFile("dummmy", text)
          val element = PsiTreeUtil.findChildOfType(psiFile, classOf[PsiErrorElement])
          if (element != null) {
            throw new RuntimeException(element.getErrorDescription)
          }

          val children = psiFile.getChildren
          val ins = children.flatMap {
            case im: EDQLImportop => {
              Seq(ImportInstruction(new URL(im.getDoubleQuotedString.getText)))
            }
            case ti: EDQLTimeoutExpr => {
              Seq(TimeoutInstruction(ti.getNumber.getText.toInt))
            }
            case hi: EDQLHeadExpr => {
              throw new RuntimeException("unsupported instruction: " + hi.getText)
            }
            case ex: EDQLExpr => parseExpr(ex)
            case w: PsiWhiteSpace => Seq()
            case c: PsiComment => Seq()
            case j => {
              throw new RuntimeException("unsupported instruction: " + j.getText)
            }
          }.toList
          scala.util.Success(ins)
        } catch {
          case ex: Throwable => scala.util.Failure(ex)
        }
      }
    })
  }


  private def parseExpr(expr: EDQLExpr): Seq[Instruction2] = {
    if (expr.getActionExpr != null) {
      return Seq(parseAction(expr.getActionExpr))
    }

    if (expr.getComment != null) {
      return Seq()
    }

    if (expr.getFunctionExpr != null) {
      val fun = expr.getFunctionExpr
      val ps = fun.getParams match {
        case null => Seq()
        case p =>
          p.getParamList.asScala.map(i => {
            i.getIdentifier0.getText
          }).toSeq
      }
      val bs = fun.getFunctionBody.getExprList.asScala.flatMap(i => parseExpr(i)).toSeq
      fun.getFunctionBody.getReturnExpr match {
        case null =>
          return Seq(FunctionInstruction(fun.getIdentifier0.getText, ps, bs))
        case r =>
          val rs = ReturnInstruction(toJsonVal(r))
          return Seq(FunctionInstruction(fun.getIdentifier0.getText, ps, bs :+ rs))
      }
    }

    if (expr.getForExpr != null) {
      val vname = expr.getForExpr.getIdentifier0.getText
      val v = toJsonVal(expr.getForExpr.getExpr)
      val bs = expr.getForExpr.getFunctionBody.getExprList.asScala.flatMap(i => parseExpr(i)).toSeq
      return Seq(ForInstruction(vname, v, bs))
    }

    if (expr.getFunctionInvokeExpr != null) {
      val funcName = expr.getFunctionInvokeExpr.getIdentifier0.getText
      val ins = expr.getFunctionInvokeExpr.getExprList.asScala.map(i => toJsonVal(i)).toSeq

      val mapIter = expr.getFunctionInvokeExpr.getMapIter
      if (mapIter != null) {
        return Seq(FunctionInvokeInstruction(funcName, ins, Some(buildMapIterInstruction(JsonCollection.Arr(), mapIter))))
      }

      return Seq(FunctionInvokeInstruction(funcName, ins))
    }

    if (expr.getOutervar != null) {
      val outervar = expr.getOutervar
      val varName = outervar.getBind.getIdentifier0.getText

      val mapIterInstruction: Option[MapIterInstruction] = if (outervar.getBind.getMapIter != null) {
        Some(buildMapIterInstruction(JsonCollection.Arr(), outervar.getBind.getMapIter))
      } else None


      if (CollectionUtils.isNotEmpty(outervar.getBind.getBinsuffixList)) {
        val j = toJsonVal(outervar.getBind)
        return Seq(VariableInstruction(varName, j))
      }

      val anonymousFun = "anonymousFun_" + RandomStringUtils.randomAlphabetic(10)

      val e = outervar.getBind.getExpr
      val v = Try.apply(toJsonVal(e))
      if (v.isSuccess) {
        v.get match {
          case arr: JsonCollection.Arr if mapIterInstruction.isDefined =>
            return Seq(
              VariableInstruction(varName, JsonCollection.Fun((anonymousFun, Seq()))),
              FunctionInstruction(anonymousFun, Seq(), Seq(mapIterInstruction.get.copy(a = arr)))
            )
          case v: JsonCollection.Var if mapIterInstruction.isDefined =>
            return Seq(
              VariableInstruction(varName, JsonCollection.Fun((anonymousFun, Seq()))),
              FunctionInstruction(anonymousFun, Seq(), Seq(mapIterInstruction.get.copy(a = v)))
            )
          case f: JsonCollection.Fun =>
            val iter = e.getFunctionInvokeExpr.getMapIter
            if (iter != null) {
              val instruction = buildMapIterInstruction(JsonCollection.Arr(), iter)
              return Seq(
                VariableInstruction(varName, JsonCollection.Fun((anonymousFun, Seq()))),
                FunctionInstruction(anonymousFun, Seq(), Seq(instruction.copy(a = f)))
              )
            }
            return Seq(VariableInstruction(varName, v.get))
          case _ =>
            return Seq(VariableInstruction(varName, v.get))
        }
      }


      return Seq(
        VariableInstruction(varName, JsonCollection.Fun((anonymousFun, Seq()))),
        FunctionInstruction(anonymousFun, Seq(), parseExpr(outervar.getBind.getExpr)))
    }


    if (expr.getArr != null && expr.getArr.getMapIter != null) {
      val arr = expr.getArr.getExprList.asScala.map(i => toJsonVal(i)).toSeq
      return Seq(buildMapIterInstruction(JsonCollection.Arr(arr: _*), expr.getArr.getMapIter))
    }

    throw new RuntimeException("unsupported instruction: " + expr.getText)
  }

  private def buildMapIterInstruction(arr: JsonCollection.Arr, mapIter: EDQLMapIter): MapIterInstruction = {
    val anonymousFun = "anonymousFun_" + RandomStringUtils.randomAlphabetic(10)
    val bs: Seq[Instruction2] = mapIter.getExprList.asScala.flatMap(i => parseExpr(i)).toSeq

    mapIter.getReturnExpr match {
      case null =>
        MapIterInstruction(arr, FunctionInstruction(anonymousFun, Seq("it"), bs))
      case r =>
        val rs = ReturnInstruction(toJsonVal(r))
        MapIterInstruction(arr, FunctionInstruction(anonymousFun, Seq("it"), bs :+ rs))
    }
  }

  private def parseAction(expr: EDQLActionExpr): Instruction2 = {
    val v = toJsonVal(expr.getObjList)

    expr.getMethod.getFirstChild.getNode.getElementType match {
      case EDQLTypes.POST => {
        PostActionInstruction(expr.getPath.getText + Option.apply(expr.getQuery).map(i => i.getText).getOrElse(""), v)
      }
      case EDQLTypes.GET => {
        GetActionInstruction(expr.getPath.getText + Option.apply(expr.getQuery).map(i => i.getText).getOrElse(""), v.headOption)
      }
      case EDQLTypes.PUT => {
        PutActionInstruction(expr.getPath.getText + Option.apply(expr.getQuery).map(i => i.getText).getOrElse(""), v.headOption)
      }
      case EDQLTypes.DELETE => {
        DeleteActionInstruction(expr.getPath.getText + Option.apply(expr.getQuery).map(i => i.getText).getOrElse(""), v.headOption)
      }
      case EDQLTypes.HEAD => {
        HeadActionInstruction(expr.getPath.getText + Option.apply(expr.getQuery).map(i => i.getText).getOrElse(""), v.headOption)
      }
      case _ => null
    }
  }

}

File: src/main/java/com/github/chengpohi/edql/parser/EDQLPsiInterceptor.kt
package com.github.chengpohi.edql.parser

import com.github.chengpohi.edql.parser.json.JsonCollection
import com.github.chengpohi.edql.parser.json.JsonValParser
import com.github.chengpohi.edql.parser.psi.*
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.util.Computable
import com.intellij.psi.PsiComment
import com.intellij.psi.PsiErrorElement
import com.intellij.psi.PsiFile
import com.intellij.psi.PsiWhiteSpace
import com.intellij.psi.util.PsiTreeUtil
import org.apache.commons.collections4.CollectionUtils
import org.apache.commons.lang3.RandomStringUtils
import java.net.URL

class EDQLPsiInterceptor(val parserFactory: EDQLParserFactory) : JsonValParser {
    fun parseJson(text: String): Result<JsonCollection.Val> {
        return ApplicationManager.getApplication().runReadAction(Computable {
            try {
                val psiFile = parserFactory.createFile("dummy", text)
                
                val element = PsiTreeUtil.findChildOfType(psiFile, PsiErrorElement::class.java)
                if (element != null) {
                    throw RuntimeException(element.errorDescription)
                }
                
                val expr = PsiTreeUtil.findChildOfType(psiFile, EDQLExpr::class.java)
                if (expr == null || !expr.textMatches(text)) {
                    throw RuntimeException("parse failed: $text")
                }
                Result.success(toJsonVal(expr))
            } catch (ex: Throwable) {
                Result.failure(ex)
            }
        })
    }
    
    fun parseExpr(text: String): Result<PsiFile> {
        return ApplicationManager.getApplication().runReadAction(Computable {
            try {
                val psiFile = parserFactory.createFile("dummy", text)
                val element = PsiTreeUtil.findChildOfType(psiFile, PsiErrorElement::class.java)
                if (element != null) {
                    throw RuntimeException(element.errorDescription)
                }
                return@Computable Result.success(psiFile)
            } catch (ex: Exception) {
                return@Computable Result.failure(ex)
            }
        })
    }
    
    fun parse(text: String): Result<List<InterceptFunction.Instruction2>> {
        return ApplicationManager.getApplication().runReadAction(Computable {
            try {
                val psiFile = parserFactory.createFile("dummy", text)
                val element = PsiTreeUtil.findChildOfType(psiFile, PsiErrorElement::class.java)
                if (element != null) {
                    throw RuntimeException(element.errorDescription)
                }
                
                val children = psiFile.children
                val ins = children.flatMap { child ->
                    when (child) {
                        is EDQLImportop -> {
                            listOf(InterceptFunction.ImportInstruction(URL(child.doubleQuotedString.text)))
                        }
                        is EDQLTimeoutExpr -> {
                            listOf(InterceptFunction.TimeoutInstruction(child.number.text.toInt()))
                        }
                        is EDQLHeadExpr -> {
                            throw RuntimeException("unsupported instruction: ${child.text}")
                        }
                        is EDQLExpr -> parseExpr(child)
                        is PsiWhiteSpace -> emptyList()
                        is PsiComment -> emptyList()
                        else -> {
                            throw RuntimeException("unsupported instruction: ${child.text}")
                        }
                    }
                }
                return@Computable Result.success(ins)
            } catch (ex: Throwable) {
                return@Computable Result.failure(ex)
            }
        })
    }
    
    private fun parseExpr(expr: EDQLExpr): List<InterceptFunction.Instruction2> {
        if (expr.actionExpr != null) {
            return listOf(parseAction(expr.actionExpr!!))
        }
        
        if (expr.comment != null) {
            return emptyList()
        }
        
        if (expr.functionExpr != null) {
            val f = expr.functionExpr!!
            val ps = f.params?.paramList?.map { i ->
                i.identifier0.text
            } ?: emptyList()
            
            val bs = f.functionBody.exprList.flatMap { i -> parseExpr(i) }
            
            return when (val returnExpr = f.functionBody.returnExpr) {
                null -> listOf(InterceptFunction.FunctionInstruction(f.identifier0.text, ps, bs))
                else -> {
                    val rs = InterceptFunction.ReturnInstruction(toJsonVal(returnExpr))
                    listOf(InterceptFunction.FunctionInstruction(f.identifier0.text, ps, bs + rs))
                }
            }
        }
        
        if (expr.forExpr != null) {
            val vname = expr.forExpr!!.identifier0.text
            val v = toJsonVal(expr.forExpr!!.expr)
            val bs = expr.forExpr!!.functionBody.exprList.flatMap { i -> parseExpr(i) }
            return listOf(InterceptFunction.ForInstruction(vname, v, bs))
        }

        if (expr.functionInvokeExpr != null) {
            val funcName = expr.functionInvokeExpr!!.identifier0.text
            val ins = expr.functionInvokeExpr!!.argList.map { it -> toJsonVal(it) }
            
            val mapIter = expr.functionInvokeExpr!!.mapIter
            if (mapIter != null) {
                return listOf(InterceptFunction.FunctionInvokeInstruction(
                    funcName, 
                    ins, 
                    buildMapIterInstruction(JsonCollection.Arr(), mapIter)
                ))
            }
            
            return listOf(InterceptFunction.FunctionInvokeInstruction(funcName, ins))
        }
        
        if (expr.outervar != null) {
            val outervar = expr.outervar!!
            val varName = outervar.bind.identifier0.text
            
            val mapIterInstruction: InterceptFunction.MapIterInstruction? = if (outervar.bind.mapIter != null) {
                buildMapIterInstruction(JsonCollection.Arr(), outervar.bind.mapIter!!)
            } else null
            
            if (CollectionUtils.isNotEmpty(outervar.bind.binsuffixList)) {
                val j = toJsonVal(outervar.bind)
                return listOf(InterceptFunction.VariableInstruction(varName, j))
            }
            
            val anonymousFun = "anonymousFun_" + RandomStringUtils.randomAlphabetic(10)
            
            val e = outervar.bind.expr
            val v = try { toJsonVal(e) } catch (ex: Throwable) {null}
            if (v != null) {
                return when (v) {
                    is JsonCollection.Arr -> {
                        if (mapIterInstruction != null) {
                            listOf(
                                InterceptFunction.VariableInstruction(varName, JsonCollection.Fun(anonymousFun to emptyList())),
                                InterceptFunction.FunctionInstruction(anonymousFun, emptyList(), listOf(mapIterInstruction.copy(a = v)))
                            )
                        } else {
                            listOf(InterceptFunction.VariableInstruction(varName, v))
                        }
                    }
                    is JsonCollection.Var -> {
                        if (mapIterInstruction != null) {
                            listOf(
                                InterceptFunction.VariableInstruction(varName, JsonCollection.Fun(anonymousFun to emptyList())),
                                InterceptFunction.FunctionInstruction(anonymousFun, emptyList(), listOf(mapIterInstruction.copy(a = v)))
                            )
                        } else {
                            listOf(InterceptFunction.VariableInstruction(varName, v))
                        }
                    }
                    is JsonCollection.Fun -> {
                        val iter = e.functionInvokeExpr?.mapIter
                        if (iter != null) {
                            val instruction = buildMapIterInstruction(JsonCollection.Arr(), iter)
                            listOf(
                                InterceptFunction.VariableInstruction(varName, JsonCollection.Fun(anonymousFun to emptyList())),
                                InterceptFunction.FunctionInstruction(anonymousFun, emptyList(), listOf(instruction.copy(a = v)))
                            )
                        } else {
                            listOf(InterceptFunction.VariableInstruction(varName, v))
                        }
                    }
                    else -> listOf(InterceptFunction.VariableInstruction(varName, v))
                }
            }
            
            return listOf(
                InterceptFunction.VariableInstruction(varName, JsonCollection.Fun(anonymousFun to emptyList())),
                InterceptFunction.FunctionInstruction(anonymousFun, emptyList(), parseExpr(outervar.bind.expr))
            )
        }
        
        if (expr.arr != null && expr.arr!!.mapIter != null) {
            val arr = expr.arr!!.exprList.map { i -> toJsonVal(i) }
            return listOf(buildMapIterInstruction(JsonCollection.Arr(*arr.toTypedArray()), expr.arr!!.mapIter!!))
        }
        
        throw RuntimeException("unsupported instruction: ${expr.text}")
    }
    
    private fun buildMapIterInstruction(arr: JsonCollection.Arr, mapIter: EDQLMapIter): InterceptFunction.MapIterInstruction {
        val anonymousFun = "anonymousFun_" + RandomStringUtils.randomAlphabetic(10)
        val bs: List<InterceptFunction.Instruction2> = mapIter.mapExpr.exprList.flatMap { i -> parseExpr(i) }
        
        return when (val returnExpr = mapIter.mapExpr.returnExpr) {
            null -> InterceptFunction.MapIterInstruction(arr, InterceptFunction.FunctionInstruction(anonymousFun, listOf("it"), bs))
            else -> {
                val rs = InterceptFunction.ReturnInstruction(toJsonVal(returnExpr))
                InterceptFunction.MapIterInstruction(arr, InterceptFunction.FunctionInstruction(anonymousFun, listOf("it"), bs + rs))
            }
        }
    }
    
    private fun parseAction(expr: EDQLActionExpr): InterceptFunction.Instruction2 {
        val v = toJsonVal(expr.objList)
        
        return when (expr.method.firstChild.node.elementType) {
            EDQLTypes.POST -> {
                InterceptFunction.PostActionInstruction(
                    expr.path.text + (expr.query?.text ?: ""), 
                    v
                )
            }
            EDQLTypes.GET -> {
                InterceptFunction.GetActionInstruction(
                    expr.path.text + (expr.query?.text ?: ""), 
                    v.firstOrNull()
                )
            }
            EDQLTypes.PUT -> {
                InterceptFunction.PutActionInstruction(
                    expr.path.text + (expr.query?.text ?: ""), 
                    v
                )
            }
            EDQLTypes.DELETE -> {
                InterceptFunction.DeleteActionInstruction(
                    expr.path.text + (expr.query?.text ?: ""), 
                    v.firstOrNull()
                )
            }
            EDQLTypes.HEAD -> {
                InterceptFunction.HeadActionInstruction(
                    expr.path.text + (expr.query?.text ?: ""), 
                    v.firstOrNull()
                )
            }
            else -> throw RuntimeException("Unsupported HTTP method: ${expr.method.text}")
        }
    }
} 